accounting/job_archive_interface.py

```diff
@@ -218,3 +220,129 @@ def append_to_where(where_stmt, conditional):
             write_records_to_file(job_records, output_file)
 
     return job_records
+
+
+def apply_decay_factor(decay_factor, acct_conn, user=None, bank=None):
+    past_usage_factors = []
+    past_usage_factors_w_decay = []
+
+    select_stmt = "SELECT * from job_usage_factor_table WHERE username=? AND bank=?"
+    dataframe = pd.read_sql_query(select_stmt, acct_conn, params=(user, bank,))
+
+    for val in dataframe.iloc[0].values:
+        if isinstance(val, float):
+            past_usage_factors.append(val)
+
+    # apply decay factor to past usage periods of a user's jobs
+    power = 1
+    for usage_factor in past_usage_factors:
+        past_usage_factors_w_decay.append(usage_factor * math.pow(decay_factor, power))
+        power += 1
+
+    # update job_usage_factor_table with new values, starting with period-2;
+    # the last usage factor in the table will get discarded after the update
+    period = 1
+    for usage_factor in past_usage_factors_w_decay[
+        : len(past_usage_factors_w_decay) - 1
+    ]:
+        update_stmt = (
+            "UPDATE job_usage_factor_table SET usage_factor_period_"
+            + str(period)
+            + "="
+            + str(usage_factor)
+            + " WHERE username='"
+            + user
+            + "' AND bank='"
+            + bank
+            + "'"
+        )
+        acct_conn.execute(update_stmt)
+        acct_conn.commit()
+        period += 1
+
+    return sum(past_usage_factors_w_decay)
+
+
+def calc_usage_factor(
+    jobs_conn,
+    acct_conn,
+    user,
+    bank,
+    priority_decay_half_life=None,
+    priority_usage_reset_period=None,
+):
+
+    # convert PriorityDecayHalfLife to a number of seconds
+    # that determine how far back to go in the job-archive
+    after_start_time = priority_decay_half_life * 604800
+
+    # get the total number of nodes and time elapsed across
+    # all of a user's jobs that completed before an end time;
+    # just store job records in a temporary .csv file that gets
+    # deleted when the calculation are done.
+    output_file = "tmp.csv"
```

Comment:
Can we avoid writing these files to a CSV?  Are the values too large to fit into memory?

Reply:
Ah, so this is because of the way `view_job_records()` works. If no `output_file` is specified, then the function defaults to printing out the job records to the screen. Do you think this is OK behavior to have when running `calc_usage_factor()`? Maybe we should change `view_job_records()` to only print out records if a certain flag is passed?